src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2999,23 +2999,53 @@ private void VisitStatementsWithLocalFunctions(BoundBlock block)
             if (!TrackingRegions && !block.LocalFunctions.IsDefaultOrEmpty)
             {
                 // First visit everything else
+                var queue1 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
                 foreach (var stmt in block.Statements)
                 {
-                    if (stmt.Kind != BoundKind.LocalFunctionStatement)
+                    if (stmt is BoundLocalFunctionStatement localFunc)
+                    {
+                        queue1.Add(localFunc);
+                    }
+                    else
                     {
                         VisitStatement(stmt);
                     }
                 }
 
-                // Now visit the local function bodies
-                foreach (var stmt in block.Statements)
+                // Now visit the local function bodies.
+                // We first visit those we have seen called (hence we know their starting state),
+                // repeating this while seeing more call sites inside other local functions,
+                // and only then we visit the completely unreachable bodies.
+                // This avoids unnecessary passes and incorrect starting states.
+                var queue2 = ArrayBuilder<BoundLocalFunctionStatement>.GetInstance();
+                for (int c = 0; c != LocalFuncUsageCount;)
```

Comment:
>c [](http://example.com/codeflow?start=25&length=1)

It looks like this local isn't used as an index. Consider giving it a meaningful name. For example, I am not sure what is its purpose.